File: python/tree/validate-binary-search-tree[1].py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def __init__(self):
        self.lastnode = None

    def isValidBST(self, root):
        if root is None:
            return True

        if not self.isValidBST(root.left):
            return False

        if self.lastnode != None and self.lastnode.val >= root.val:
            return False

        self.lastnode = root
        return self.isValidBST(root.right)

File: c/tree/validate-binary-search-tree[1].c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode *lastnode;
bool isValid(struct TreeNode* root)
{
    if(root == NULL)
        return true;

    if(!isValid(root->left))
        return false;

    if(lastnode != NULL && lastnode->val >= root->val)
        return false;

    lastnode = root;
    return isValid(root->right);
}

bool isValidBST(struct TreeNode* root) {
    lastnode = NULL;
    return isValid(root);
}
